File: go/schemapb/bake.go
package schemapb

import (
	"errors"
	"fmt"

	"google.golang.org/protobuf/proto"
)

// Bake validates and resolves values, then seals them with the schema into a
// Baked snapshot (canonical wire values). On a blocking (ERROR) failure baked
// is nil; warnings do not block and are returned alongside the Baked. The
// returned error is programmatic only (schema does not compile, value cannot
// be canonicalised).
func (s *Schema) Bake(values map[string]any) (*Baked, *ValidationResult, error) {
	e, err := s.engine()
	if err != nil {
		return nil, nil, err
	}
	return e.Bake(values)
}

// Bake is the compiled-engine form of (*Schema).Bake.
func (e *Engine) Bake(values map[string]any) (*Baked, *ValidationResult, error) {
	res := e.Validate(values) // resolves values in place + checks
	if res.Blocking() {
		return nil, res, nil
	}
	st, err := e.canonicalStruct(values)
	if err != nil {
		return nil, res, fmt.Errorf("schemapb: bake: %w", err)
	}
	return &Baked{Schema: e.schema, Values: st}, res, nil
}

// canonicalStruct projects a resolved native form into the contract's
// canonical wire variants, keyed by the schema's declared fields (unknown
// keys fall back to best-fit conversion).
func (e *Engine) canonicalStruct(values map[string]any) (*StructValue, error) {
	fields := make(map[string]*Value, len(values))
	for name, val := range values {
		f := findField(e.schema.GetFields(), name)
		var v *Value
		var err error
		switch {
		case f == nil:
			v, err = FromGo(val)
		case f.GetRef() != nil:
			if def := e.schema.GetDefs()[refDefKey(f.GetRef())]; def != nil {
				if m, ok := val.(map[string]any); ok {
					v, err = canonicalStruct(def, m, name)
					break
				}
			}
			v, err = FromGo(val)
		case f.GetOneOf() != nil:
			if variant, m := selectVariant(f.GetOneOf(), val); variant != nil {
				v, err = canonicalStruct(variant, m, name)
				break
			}
			v, err = FromGo(val)
		default:
			v, err = CanonicalValue(f, val)
		}
		if err != nil {
			return nil, err
		}
		fields[name] = v
	}
	return &StructValue{Fields: fields}, nil
}

// Merge layers overrides onto a baked form and re-seals against the same
// schema. Nested objects merge recursively; lists append unless replaceLists;
// immutable fields keep their baked values (a changed immutable is rejected).
// It compiles the embedded schema with default options; a schema needing
// custom compile options (format extensions) must merge through its own
// engine — see (*Engine).Merge.
func (b *Baked) Merge(overrides *StructValue, replaceLists bool) (*Baked, *ValidationResult, error) {
	base := b.GetValues().ToGo()
	ov := overrides.ToGo()
	return b.GetSchema().Bake(mergeMaps(base, ov, replaceLists))
}

// Merge is (*Baked).Merge evaluated on this engine (keeping its compile
// options, e.g. WithFormats extensions).
func (e *Engine) Merge(b *Baked, overrides *StructValue, replaceLists bool) (*Baked, *ValidationResult, error) {
	base := b.GetValues().ToGo()
	ov := overrides.ToGo()
	return e.Bake(mergeMaps(base, ov, replaceLists))
}

// Matches reports whether the baked schema is identical in content to s.
func (b *Baked) Matches(s *Schema) bool {
	return proto.Equal(b.GetSchema(), s)
}

// Bake validates and seals an inline Filled into a Baked. It works only for a
// Filled carrying an inline schema; a Filled that references a schema by id
// must be baked where a registry can resolve the id.
func (f *Filled) Bake() (*Baked, *ValidationResult, error) {
	s := f.GetSchema().GetSchema()
	if s == nil {
		return nil, nil, errors.New("schemapb: Filled.Bake requires an inline schema (id refs resolve via a registry)")
	}
	return s.Bake(f.GetValues().ToGo())
}

// mergeMaps deep-merges src over dst (objects recurse; lists append unless
// replaceLists; scalars overwrite). It returns a new map.
func mergeMaps(dst, src map[string]any, replaceLists bool) map[string]any {
	out := make(map[string]any, len(dst))
	for k, v := range dst {
		out[k] = v
	}
	for k, sv := range src {
		if dv, ok := out[k]; ok {
			if dm, ok1 := dv.(map[string]any); ok1 {
				if sm, ok2 := sv.(map[string]any); ok2 {
					out[k] = mergeMaps(dm, sm, replaceLists)
					continue
				}
			}
			if !replaceLists {
				if dl, ok1 := dv.([]any); ok1 {
					if sl, ok2 := sv.([]any); ok2 {
						out[k] = append(append([]any{}, dl...), sl...)
						continue
					}
				}
			}
		}
		out[k] = sv
	}
	return out
}

File: py/src/schemapb/bake.py
"""Bake / merge / render, mirroring the Go reference bake.go + render.go."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING

import chevron

from schemapb._gen.schemapb import (
    Baked,
    Filled,
    Schema,
    SchemaField,
    StructValue,
    ValidationResult,
    Value,
)
from schemapb.compute import field_is_active, ref_def_key, select_variant
from schemapb.render import RenderField, display_string, render_field
from schemapb.validate import result_blocking, validate
from schemapb.value import (
    CanonicalError,
    Native,
    NativeStruct,
    canonical_struct,
    canonical_value,
    from_native,
    struct_to_native,
)

if TYPE_CHECKING:
    from schemapb.engine import Engine


@dataclass(slots=True)
class BakeOutcome:
    result: ValidationResult
    baked: Baked | None = None


def bake(e: Engine, values: NativeStruct) -> BakeOutcome:
    """Validate + resolve, then seal in canonical wire form."""
    result = validate(e, values)
    if result_blocking(result):
        return BakeOutcome(result=result)
    return BakeOutcome(
        result=result,
        baked=Baked(schema=e.schema, values=_canonical_engine_struct(e, values)),
    )


def _canonical_engine_struct(e: Engine, values: NativeStruct) -> StructValue:
    by_name = {f.name: f for f in e.schema.fields}
    fields = {name: _canonical_top(e, by_name.get(name), val) for name, val in values.items()}
    return StructValue(fields=fields)


def _canonical_top(e: Engine, f: SchemaField | None, val: Native) -> Value:
    if f is None:
        return from_native(val)
    if f.ref is not None:
        def_ = e.schema.defs.get(ref_def_key(f.ref))
        if def_ is not None and isinstance(val, dict):
            return canonical_struct(def_, val)
        return from_native(val)
    if f.one_of is not None:
        sel = select_variant(f.one_of, val)
        if sel is not None:
            return canonical_struct(sel[0], sel[1])
        return from_native(val)
    try:
        return canonical_value(f, val)
    except CanonicalError:
        return from_native(val)


def merge(e: Engine, baked: Baked, overrides: StructValue, *, replace_lists: bool) -> BakeOutcome:
    base = struct_to_native(baked.values)
    over = struct_to_native(overrides)
    return bake(e, _merge_structs(base, over, replace_lists=replace_lists))


def _merge_structs(dst: NativeStruct, src: NativeStruct, *, replace_lists: bool) -> NativeStruct:
    out: NativeStruct = dict(dst)
    for k, sv in src.items():
        dv = out.get(k)
        if isinstance(dv, dict) and isinstance(sv, dict):
            out[k] = _merge_structs(dv, sv, replace_lists=replace_lists)
            continue
        if not replace_lists and isinstance(dv, list) and isinstance(sv, list):
            out[k] = [*dv, *sv]
            continue
        out[k] = sv
    return out


def baked_matches(baked: Baked, s: Schema) -> bool:
    return baked.schema is not None and baked.schema == s


def filled_bake(filled: Filled, compile_fn: Callable[[Schema], Engine]) -> BakeOutcome:
    src = filled.schema
    if src is None or src.schema is None:
        msg = "schemapb: Filled bake requires an inline schema (id refs resolve via a registry)"
        raise ValueError(msg)
    return bake(compile_fn(src.schema), struct_to_native(filled.values))


def render(e: Engine, name: str, values: NativeStruct) -> str | None:
    tmpl = e.schema.templates.get(name)
    if tmpl is None:
        return None
    return str(chevron.render(tmpl, build_render_context(e, values)))


def render_baked(e: Engine, baked: Baked, name: str) -> str | None:
    return render(e, name, struct_to_native(baked.values))


def build_render_context(e: Engine, values: NativeStruct) -> dict[str, object]:
    """The contract render context; inactive fields excluded entirely."""
    fields: list[RenderField] = []
    group_fields: list[list[RenderField]] = []
    group_names: list[str] = []
    group_idx: dict[str, int] = {}

    for f in e.schema.fields:
        if (f.when or "") != "" and not field_is_active(e, f, values, f.name, None):
            continue
        rf = render_field(f, values)
        fields.append(rf)
        g = f.group or ""
        i = group_idx.get(g)
        if i is None:
            i = len(group_names)
            group_idx[g] = i
            group_names.append(g)
            group_fields.append([])
        group_fields[i].append(rf)
    groups: list[dict[str, object]] = [
        {"name": name, "fields": flds} for name, flds in zip(group_names, group_fields, strict=True)
    ]

    display = {name: "" if val is None else display_string(val) for name, val in values.items()}
    return {"fields": fields, "groups": groups, "values": display}
